pkg/reduce/readloop/readloop.go

```diff
@@ -241,16 +238,18 @@ func (rl *ReadLoop) upsertWindowsAndKeys(m *isb.ReadMessage) []*keyed.KeyedWindo
 		return []*keyed.KeyedWindow{}
 	}
 
-	processingWindows := rl.windowingStrategy.AssignWindow(m.EventTime)
+	processingWindows := rl.windower.AssignWindow(m.EventTime)
 	var kWindows []*keyed.KeyedWindow
 	for _, win := range processingWindows {
-		kw := rl.aw.GetKeyedWindow(win)
-		if kw == nil {
-			kw = rl.aw.CreateKeyedWindow(win)
-			rl.log.Debugw("Creating new keyed window", zap.Any("key", kw.Keys), zap.Int64("startTime", kw.Start.UnixMilli()), zap.Int64("endTime", kw.End.UnixMilli()))
+		w := rl.windower.GetWindow(win)
+		if w == nil {
+			w = rl.windower.CreateWindow(win)
 		}
 		// track the key to window relationship
+		kw := w.(*keyed.KeyedWindow)
```

Comment:
looks like we always need to convert an AlignedWindow to KeyedWindow, is there a way to do this more naturally without type conversion?

Reply:
we dont have key information in aligned window and hence the cast.